proposals/0642-JavaScriptCore-safe-bundle-URLs.md

```diff
@@ -0,0 +1,64 @@
+
+`http://localhost:8080/project/index.bundle;&platform=ios`
+
+Should be supported by Metro (and other bundlers) equivalently and in addition to the current query-string form:
+
+`http://localhost:8080/project/index.bundle?platform=ios`
+
+## Motivation
+
+Recent versions of JavaScriptCore strip query strings from source URLs in stack traces (`Error.prototype.stack`).
+
+To implement features such as Metro's `/symbolicate` endpoint, we must be able to derive (a close approximation of) the source that was executed the client. To do so correctly we need to know the build parameters the client used originally. Those parameters being stripped (and defaults used instead) has manifested as obscure resolution errors for React Native users, on both Expo and the community CLI ([1](https://github.com/facebook/react-native/issues/36794), [2](https://github.com/expo/expo/issues/22008)).
+
+Ideally, for maximal compatibility with debuggers (etc.) that may intercept the raw error stack, the source URLs in the error stack should be sufficient, and should be valid URLs, without any post-processing or external context.
+
+## Detailed design
+
+Use a character sequence from the "reserved" set specified in RFC 3986, namely `;&`, as an alternative delimiter in place of `?`, and percent-encode any `?` characters appearing within the subsequent query string, so that the result is a valid [`path`](https://www.rfc-editor.org/rfc/rfc3986#section-3.3).
+
+If the client sends a URL in this format, Metro should respond as if `?` was seen in place of the alternative delimiter, with the following exceptions:
+ - Wherever the bundle itself encodes its own URL, source map URL, or the URLs of other bundles on the same server (such as for lazy bundling), those should use the alternative format for any included query strings *if and only if* the original bundle request used the alternative format. Otherwise, standard query strings starting with `?` should be emitted.
+ - URLs of mixed format with `;&` preceding `?`, i.e. `.../path;&foo=bar?foo=baz&bar=quux` should be interpreted leniently, as if they were given fully in the alternative format, i.e. `.../path;&foo=bar&foo=baz&bar=quux`.
+
+Because the alternative delimiter is made of reserved characters, it does not create ambiguity with standard URLs and may be escaped by percent-encoding for use as a param value, path component, etc. In particular, this supports the edge case of a valid POSIX file path containing the substring `;&` being the subject of a bundle request.
+
+## Drawbacks
+
+ - The proposed URL format is unfamiliar, and standard libraries and tools won't be able to encode/decode build parameters into/from URLs in this format. An exposed API or package would help mitigate.
+ - Maintenance/complexity of supporting an additional URL format.
+ - Forking of bundle URL generation in React Native by platform/engine, assuming we use this new format on as narrow a range of target engines as possible.
+
+## Alternatives
```

Comment:
- What prevents Safari from also stripping out `;&` for security reasons?
- Have you considered embedding a comment similar to `//# sourceMappingURL=` for source maps?

Reply:
> - What prevents Safari from also stripping out ;& for security reasons?
 
Nothing in theory, but my understanding is that query strings (which of course aren't fundamentally less secure than path components) are commonly used to append tracking or other sensitive information in a web context during redirection. From what I can gather (though the security bugzilla is private), that was to plug quite a specific hole where stack consumers could end up being sent the sensitive data added by a separate redirection service. Those services would have to motivation to make that private data leakable, and couldn't append `;&` in the general case anyway because it will usually break routing.

I've also been in touch with Apple folks about building an opt-out for controlled, non-browser use cases like RN - they're open to that because the disruption to our use case was completely unintentional, so I don't think we're getting into an arms race here - just a workaround for a temporary situation.
 
> - Have you considered embedding a comment similar to //# sourceMappingURL= for source maps?

Metro already adds these, but JavaScriptCore doesn't use them when populating error stacks. The stacks coming from the engine only use the `sourceURL` provided explicitly to the `JSEvaluateScript` call ([docs](https://developer.apple.com/documentation/javascriptcore/1451589-jsevaluatescript)).